compiler/passes/buildDefaultFunctions.cpp

```diff
@@ -811,44 +820,142 @@ static void buildRecordComparisonFunc(AggregateType* ct, const char* op) {
+  else {
+    INT_FATAL("Unrecognized operator was passed to buildRecordComparisonFunc");
+  }
+
+  DefExpr* def = new DefExpr(fn);
+  ct->symbol->defPoint->insertBefore(def);
+  reset_ast_loc(def, ct->symbol);
+  normalize(fn);
+}
+
+static void buildRecordEqualsBody(AggregateType *ct, FnSymbol *fn,
+                                  ArgSymbol *arg1, ArgSymbol *arg2) {
+  // add comparisons for fields
+  for_fields(tmp, ct) {
+    if (!tmp->hasFlag(FLAG_IMPLICIT_ALIAS_FIELD) &&
+        !tmp->hasFlag(FLAG_TYPE_VARIABLE)) {  // types fields must be equal
+      Expr* left = new CallExpr(tmp->name, gMethodToken, arg1);
+      Expr* right = new CallExpr(tmp->name, gMethodToken, arg2);
+      CallExpr *elemComp = new CallExpr("!=", left, right);
+      fn->insertAtTail(new CondStmt(elemComp,
+                                    new CallExpr(PRIM_RETURN, gFalse)));
+    }
+  }
+  // either it is an empty record, and we didn't add any meaningful comparisons,
+  // or we added those comparisons but they never returned
+
+  // for ==, in either case we return true
+  fn->insertAtTail(new CallExpr(PRIM_RETURN, gTrue));
+}
+
+static void buildRecordNotEqualsBody(AggregateType *ct, FnSymbol *fn,
+                                  ArgSymbol *arg1, ArgSymbol *arg2) {
   // add comparisons for fields
   for_fields(tmp, ct) {
     if (!tmp->hasFlag(FLAG_IMPLICIT_ALIAS_FIELD) &&
         !tmp->hasFlag(FLAG_TYPE_VARIABLE)) {  // types fields must be equal
       Expr* left = new CallExpr(tmp->name, gMethodToken, arg1);
       Expr* right = new CallExpr(tmp->name, gMethodToken, arg2);
-      CallExpr *elemComp = new CallExpr(op, left, right);
-      if (isNotEqual) {
-        fn->insertAtTail(new CondStmt(elemComp,
+      CallExpr *elemComp = new CallExpr("!=", left, right);
+      fn->insertAtTail(new CondStmt(elemComp,
+                                    new CallExpr(PRIM_RETURN, gTrue)));
+    }
+  }
+  // either it is an empty record, and we didn't add any meaningful comparisons,
+  // or we added those comparisons but they never returned
+
+  // for !=, in either case we return fales
+  fn->insertAtTail(new CallExpr(PRIM_RETURN, gFalse));
+}
+
+static void buildRecordLessThanBody(AggregateType *ct, FnSymbol *fn,
+                                    ArgSymbol *arg1, ArgSymbol *arg2,
+                                    bool allowEquals) {
+  // add comparisons for fields
+  for_fields(tmp, ct) {
+    if (!tmp->hasFlag(FLAG_IMPLICIT_ALIAS_FIELD) &&
+        !tmp->hasFlag(FLAG_TYPE_VARIABLE)) {  // types fields must be equal
+      {
+        Expr* left = new CallExpr(tmp->name, gMethodToken, arg1);
+        Expr* right = new CallExpr(tmp->name, gMethodToken, arg2);
+        CallExpr *elemCompTrue = new CallExpr("<", left, right);
+        fn->insertAtTail(new CondStmt(elemCompTrue,
                                       new CallExpr(PRIM_RETURN, gTrue)));
       }
-      else {
-        fn->insertAtTail(new CondStmt(new CallExpr("!", elemComp), 
+      {
```

Comment:
This doesn't look like an if/else branch?  You've got two `{` after the if statement, and then you close one and open another one?  Was that intentional?

Reply:
I'd rather they were just two code blocks separated only by line breaks, personally